### PR TITLE
React app

### PR DESCRIPTION
**Pull request by Vitaminvp**
### React app

[Demo](https://vitaminvp.github.io/Unsplash-react-redux-thunk-ts/) |
[Code base](https://github.com/Vitaminvp/Unsplash-react-redux-thunk-ts)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/Vitaminvp/react-app/App.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            </NavLink>
+          ))}
+        </Nav>
+        <Switch>
+          {routes.map(({ path, exact, Component, title }) => (
+            <Route
+              key={path}
+              path={path}
+              exact={exact}
+              render={routeProps => <Component {...routeProps} />}
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/App.tsx]:
```suggestion
              render={Component}
```

You don't need function

File: submissions/Vitaminvp/react-app/reducers/auth.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { Action } from "../types/action";
+import { AUTH_ACTION_TYPES } from "../actions/auth";
+
+export interface authState {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/reducers/auth.ts]: `interface`, `enum` and `type` must be called from big letter

File: submissions/Vitaminvp/react-app/Components/auth/Auth.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    );
+  }
+}
+
+const mapStateToProps = (state: State): authState => ({
+  isAuthenticated: state.auth.isAuthenticated,
+  token: state.auth.token
+});
+
+const mapDispatchToProps = (dispatch: Dispatch<AuthAction>) => {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/auth/Auth.tsx]:
```suggestion
const mapDispatchToProps = {
  setToken
}
```

https://react-redux.js.org/api/connect#object-shorthand-form

File: submissions/Vitaminvp/react-app/Components/button/Button.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+interface IProps extends ButtonHTMLAttributes<HTMLButtonElement> {
+  classNames?: string;
+  type?: ButtonTypes;
+}
+
+export const Button: React.FunctionComponent<IProps> = ({
+  children,
+  classNames,
+  ...props
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/button/Button.tsx]:
```suggestion
  ...buttonProps
```

File: submissions/Vitaminvp/react-app/Components/grid/Grid.tsx [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  total: number;
+  filterInput: string;
+  currentPage: number;
+  searchInput: string;
+  sortingParam: string;
+
+  onItemsAdded(searchInput: string, currentPage: number): void;
+}
+
+class Grid extends React.PureComponent<Props, {}> {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/grid/Grid.tsx]: No need for empty object as state.

Also, why `PureComponent`?

File: submissions/Vitaminvp/react-app/Components/grid/Grid.tsx [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  private loadImages = () => {
+    const currentPage = this.props.currentPage + 1;
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/grid/Grid.tsx]: Is it possible, that you will add page more than max?

File: submissions/Vitaminvp/react-app/Components/grid/Grid.tsx [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    this.props.onItemsAdded(this.props.searchInput, currentPage);
+  };
+
+  private handleFilter() {
+    const { items, filterInput } = this.props;
+    return [...(items as Array<Image>)].filter(item => {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/grid/Grid.tsx]: Why? You already added types for `items` in `Props`

File: submissions/Vitaminvp/react-app/Components/grid/Grid.tsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    return items.sort((a, b) => handleSorting[sortingParam](a, b));
+  }
+
+  render() {
+    const { items, total } = this.props;
+    const isButtonVisible = items && items.length > 0 && total > items.length;
+    this.handleSort();
+    return (
+      <>
+        <Suspense fallback={<div>Loading...</div>}>
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/grid/Grid.tsx]: As I already asked and looks like you have ignored my questions - it is not working right now, only with `import`. Or I mistaking?

File: submissions/Vitaminvp/react-app/Components/grid/Grid.tsx [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    });
+  }
+
+  private handleSort() {
+    const { items, sortingParam } = this.props;
+    const sorting = (
+      a: { alt_description: string },
+      b: { alt_description: string }
+    ): number => (a.alt_description > b.alt_description ? 1 : -1);
+    const handleSorting: {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/grid/Grid.tsx]: Do you need such monstrous types?

File: submissions/Vitaminvp/react-app/Components/grid/Grid.tsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    const isButtonVisible = items && items.length > 0 && total > items.length;
+    this.handleSort();
+    return (
+      <>
+        <Suspense fallback={<div>Loading...</div>}>
+          <div className={"nav__container"}>
+            <SearchForm className="nav__search-form" />
+          </div>
+          <div className={"grid"}>
+            {this.handleFilter().map(
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/grid/Grid.tsx]: I believe, TS will understand, that you returning `Image` here. So no need in this typings

File: submissions/Vitaminvp/react-app/Components/grid/Grid.tsx [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+          >
+            Show more {total ? `(${items.length} of ${total})` : ""}
+          </Button>
+        ) : null}
+      </>
+    );
+  }
+}
+
+const mapStateToProps = (state: { unsplash: Props }) => {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/grid/Grid.tsx]:
```suggestion
const mapStateToProps = (state: { unsplash: Props }) => unsplash
```

File: submissions/Vitaminvp/react-app/Components/grid/index.ts [hunk 12]
```diff
@@ -0,0 +1,5 @@
+import Grid from "./Grid";
+export { Grid };
+
+
+
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/grid/index.ts]: Redundant empty strings

File: submissions/Vitaminvp/react-app/Components/gridItem/GridItem.tsx [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  link?: string;
+  height?: number;
+  classNames?: string;
+}
+
+interface IState {
+  value: string;
+}
+
+export class GridItem extends React.Component<IProps, IState> {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/gridItem/GridItem.tsx]: Can be replaced with functional component

File: submissions/Vitaminvp/react-app/Components/image/Image.tsx [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+class Image extends React.Component<IProps> {
+  componentDidMount() {
+    this.props.onLoad(this.props.match.params.id);
+  }
+  render() {
+    if (this.props.imageItem) {
+      const { alt_description, user, urls } = this.props.imageItem;
+      return (
+        <Suspense fallback={<div>Loading...</div>}>
+          <article>
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/image/Image.tsx]: Image in `article`? I think name `Image` is not what this component is

File: submissions/Vitaminvp/react-app/Components/image/Image.tsx [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        </Suspense>
+      );
+    } else {
+      return null;
+    }
+  }
+}
+
+const mapStateToProps = (state: { unsplash: UnsplashState }) => {
+  return {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/image/Image.tsx]: Redundant return

File: submissions/Vitaminvp/react-app/Components/image/Image.tsx [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    imageItem: state.unsplash.item
+  };
+};
+
+const mapDispatchToProps = (dispatch: Dispatch) => {
+  return {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/image/Image.tsx]: same

File: submissions/Vitaminvp/react-app/Components/nav/Nav.tsx [hunk 17]
```diff
@@ -0,0 +1,19 @@
+import React, { ReactElement, ReactNode } from "react";
+import "./Nav.scss";
+
+export const Nav: React.FunctionComponent<ReactNode> = ({ children }) => {
+  return (
+    <nav className={"nav"}>
+      <ul className={"nav-menu"}>
+        {React.Children.map(
+          children as ReactElement,
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/nav/Nav.tsx]: WTF, what for?

File: submissions/Vitaminvp/react-app/Components/nav/Nav.tsx [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+          (child: React.ReactElement) => {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/nav/Nav.tsx]: Why?

File: submissions/Vitaminvp/react-app/Components/nav/Nav.tsx [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+            return (
+              <li key={child!.props.children}>{React.cloneElement(child)}</li>
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/nav/Nav.tsx]: `child!` <- remove this. Bad practice and whole in typings

File: submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+class SearchForm extends React.Component<Props, State> {
+  state = {
+    searchInput: "",
+    filterInput: "",
+    sortingParam: ""
+  };
+
+  componentDidMount(): void {
+    const { searchInput, sortingParam } = this.props;
+    if (!searchInput) this.props.onSubmit("cars", 1);
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx]: Why `1`? Why not `3.14`?

File: submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    this.setState({ sortingParam });
+  }
+
+  private onFormChange = (e: SyntheticEvent<HTMLFormElement>) => {
+    const { onChange } = this.props;
+    const target = e.target as Target;
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx]: What for?

File: submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+    const targetValue =
+      target.name === "sortingParam" ? target.id : target.value;
+    this.setState(
+      state => ({
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx]: Reason for functional `setState`?

File: submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+    return (
+      <form
+        onSubmit={this.onSubmit}
+        className={classNames}
+        onChange={this.onFormChange}
+      >
+        <div className="search">
+          <Input name="searchInput" type="text" label="Search" />
+          <Button type={ButtonTypes.SUBMIT} classNames={"btn-search"}>
+            {" "}
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx]: Empty string node? What for?

File: submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+    const classNames = classnames("search-form", "native-form");
+    return (
+      <form
+        onSubmit={this.onSubmit}
+        className={classNames}
+        onChange={this.onFormChange}
+      >
+        <div className="search">
+          <Input name="searchInput" type="text" label="Search" />
+          <Button type={ButtonTypes.SUBMIT} classNames={"btn-search"}>
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx]: No need for brackets

File: submissions/Vitaminvp/react-app/actions/filter.ts [hunk 25]
```diff
@@ -0,0 +1,13 @@
+import { ActionTypes } from "./unsplash";
+
+export const filterActionCreator = (payload: {
+  filterInput: string;
+  sortingParam: string;
+}): any => {
+  return (dispatch: any) => {
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/actions/filter.ts]: Pls, use `@types/react-redux`

File: submissions/Vitaminvp/react-app/index.tsx [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+import './index.scss';
+import App from './App';
+import {Provider} from "react-redux";
+import {createStore, applyMiddleware, compose} from "redux";
+import reducers from "./reducers";
+import reduxThunk from "redux-thunk";
+import {HashRouter} from "react-router-dom";
+
+//@ts-ignore
+const composedEnhancer = window.__REDUX_DEVTOOLS_EXTENSION__;
```
Comment (zonzujiro) [inline submissions/Vitaminvp/react-app/index.tsx]: It is not composed, actually. It is a simple middleware here

File: submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx (hunk 22 above)
Comment (Vitaminvp) [reply submissions/Vitaminvp/react-app/Components/searchForm/SearchForm.tsx]: I use new state to dispatch values, I must be sure of new values.

File: submissions/Vitaminvp/react-app/index.tsx (hunk 26 above)
Comment (Vitaminvp) [reply submissions/Vitaminvp/react-app/index.tsx]: You mean just new name "enhancer"?

**Pull request closed without merging**